packages/kit/src/exports/vite/preview/index.js

```diff
@@ -49,84 +51,143 @@ export async function preview(vite, vite_config, svelte_config) {
+		// prerendered pages (we can't just use sirv because we need to
+		// preserve the correct trailingSlash behaviour)
+		vite.middlewares.use(
+			scoped(base, (req, res, next) => {
+				let if_none_match_value = req.headers['if-none-match'];
 
-				/** @type {string | undefined} */
-				let redirect;
+				if (if_none_match_value?.startsWith('W/"')) {
+					if_none_match_value = if_none_match_value.substring(2);
+				}
 
-				if (is_file(html_filename)) {
-					filename = html_filename;
-					prerendered = true;
-				} else if (has_trailing_slash) {
-					if (is_file(filename.slice(0, -1) + '.html')) {
-						redirect = pathname.slice(0, -1);
-					}
-				} else if (is_file(filename + '/index.html')) {
-					redirect = pathname + '/';
+				if (if_none_match_value === etag) {
+					res.statusCode = 304;
+					res.end();
+					return;
 				}
 
-				if (redirect) {
-					if (search) redirect += search;
-					res.writeHead(307, {
-						location: redirect
-					});
+				const { pathname, search } = new URL(/** @type {string} */ (req.url), 'http://dummy');
+
+				let filename = normalizePath(
+					join(svelte_config.kit.outDir, 'output/prerendered/pages' + pathname)
+				);
+				let prerendered = is_file(filename);
+
+				if (!prerendered) {
+					const has_trailing_slash = pathname.endsWith('/');
+					const html_filename = `${filename}${has_trailing_slash ? 'index.html' : '.html'}`;
+
+					/** @type {string | undefined} */
+					let redirect;
+
+					if (is_file(html_filename)) {
+						filename = html_filename;
+						prerendered = true;
+					} else if (has_trailing_slash) {
+						if (is_file(filename.slice(0, -1) + '.html')) {
+							redirect = pathname.slice(0, -1);
+						}
+					} else if (is_file(filename + '/index.html')) {
+						redirect = pathname + '/';
+					}
 
-					res.end();
+					if (redirect) {
+						if (search) redirect += search;
+						res.writeHead(307, {
+							location: redirect
+						});
 
-					return;
+						res.end();
+
+						return;
+					}
 				}
-			}
 
-			if (prerendered) {
-				res.writeHead(200, {
-					'content-type': lookup(pathname) || 'text/html',
-					etag
-				});
+				if (prerendered) {
+					res.writeHead(200, {
+						'content-type': lookup(pathname) || 'text/html',
+						etag
+					});
 
-				fs.createReadStream(filename).pipe(res);
-			} else {
-				next();
-			}
-		});
+					fs.createReadStream(filename).pipe(res);
+				} else {
+					next();
+				}
+			})
+		);
 
 		// SSR
 		vite.middlewares.use(async (req, res) => {
 			const host = req.headers['host'];
-			req.url = req.originalUrl;
 
-			const request = await getRequest({
-				base: `${protocol}://${host}`,
-				request: req
-			});
+			let request;
+
+			try {
+				request = await getRequest({
+					base: `${protocol}://${host}`,
+					request: req
+				});
+			} catch (/** @type {any} */ err) {
+				res.statusCode = err.status || 400;
+				return res.end('Invalid request body');
+			}
```

Comment:
why is this `try-catch` block here? should be unnecessary post-#11289

Reply:
good catch! I went back to the 1.x version and accidentally lost that. I just restored it and verified that we're not missing any other changes